### PR TITLE
provide a way to copy process vars into task

### PR DESCRIPTION
**Pull request by ryandawsonuk**
for https://github.com/Activiti/Activiti/issues/2104

### CODE REVIEW

**Comment by ryandawsonuk** [comment]
Closing this PR for now as it is currently breaking TaskRuntimeCompleteTaskTest in the spring boot starter:

```
org.activiti.engine.ActivitiIllegalArgumentException: executionId is null
	at org.activiti.engine.impl.cmd.NeedsActiveExecutionCmd.execute(NeedsActiveExecutionCmd.java:40) ~[classes/:na]
	at org.activiti.engine.impl.interceptor.CommandInvoker$1.run(CommandInvoker.java:37) ~[classes/:na]
	at org.activiti.engine.impl.interceptor.CommandInvoker.executeOperation(CommandInvoker.java:78) ~[classes/:na]
	at org.activiti.engine.impl.interceptor.CommandInvoker.executeOperations(CommandInvoker.java:57) ~[classes/:na]
	at org.activiti.engine.impl.interceptor.CommandInvoker.execute(CommandInvoker.java:42) ~[classes/:na]
	at org.activiti.engine.impl.interceptor.TransactionContextInterceptor.execute(TransactionContextInterceptor.java:48) ~[classes/:na]
	at org.activiti.engine.impl.interceptor.CommandContextInterceptor.execute(CommandContextInterceptor.java:63) ~[classes/:na]
	at org.activiti.spring.SpringTransactionInterceptor$1.doInTransaction(SpringTransactionInterceptor.java:47) [classes/:na]
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:140) [spring-tx-5.0.8.RELEASE.jar:5.0.8.RELEASE]
	at org.activiti.spring.SpringTransactionInterceptor.execute(SpringTransactionInterceptor.java:45) [classes/:na]
	at org.activiti.engine.impl.interceptor.LogInterceptor.execute(LogInterceptor.java:29) [classes/:na]
	at org.activiti.engine.impl.cfg.CommandExecutorImpl.execute(CommandExecutorImpl.java:44) [classes/:na]
	at org.activiti.engine.impl.cfg.CommandExecutorImpl.execute(CommandExecutorImpl.java:39) [classes/:na]
	at org.activiti.engine.impl.RuntimeServiceImpl.setVariables(RuntimeServiceImpl.java:231) ~[classes/:na]
	at org.activiti.engine.impl.bpmn.helper.TaskVariableCopier.copyVariablesOutFromTaskLocal(TaskVariableCopier.java:20) ~[classes/:na]
	at org.activiti.engine.impl.cmd.CompleteTaskCmd.execute(CompleteTaskCmd.java:77) ~[classes/:na]
	at org.activiti.engine.impl.cmd.CompleteTaskCmd.execute(CompleteTaskCmd.java:26) ~[classes/:na]
```

File: activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/helper/TaskVariableCopier.java [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        task.setVariablesLocal(task.getVariables());
+
+
+    }
+
+    public static void copyVariablesOutFromTaskLocal(TaskEntity task){
+
+        //TODO: would like to filter which variables copy much as with subProcesses
+
+        //provided not a standalone task
```
Comment (salaboy) [inline activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/helper/TaskVariableCopier.java]: @ryandawsonuk is this not redundant? the Behaviour will not be called unless the task creation happens from a process instance.

File: pom.xml [hunk 2]
```diff
@@ -17,7 +17,7 @@
   <url>http://activiti.org</url>
   <properties>
     <activiti-build.version>7.0.39</activiti-build.version>
-    <activiti-api.version>7.0.37</activiti-api.version>
+    <activiti-api.version>7.0.0-SNAPSHOT</activiti-api.version>
```
Comment (salaboy) [inline pom.xml]: @ryandawsonuk ? do we need this? I think that we should keep the fixed version here.

File: activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/helper/TaskVariableCopier.java (hunk 1 above)
Comment (ryandawsonuk) [reply activiti-engine/src/main/java/org/activiti/engine/impl/bpmn/helper/TaskVariableCopier.java]: the check is for completion as that goes through CompleteTaskCmd and standalone tasks also use that

File: activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/tasks/TaskVariablesLocalCopiesIT.java [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        startVariables.put("start2","start2");
+
+        //when
+        ProcessInstance twoTaskInstance = processRuntime.start(ProcessPayloadBuilder.start()
+                .withProcessDefinitionKey(TWOTASK_PROCESS)
+                .withVariables(startVariables)
+                .build());
+
+        assertThat(processRuntime.variables(ProcessPayloadBuilder.variables().withProcessInstance(twoTaskInstance).build()))
+                .extracting("name", "value")
```
Comment (erdemedeiros) [inline activiti-spring-boot-starter/src/test/java/org/activiti/spring/boot/tasks/TaskVariablesLocalCopiesIT.java]: Prefer method references to literal method name: `.extracting(VariableInstance::getName(), VariableInstance::getValue)`.

File: activiti-engine/src/main/java/org/activiti/engine/impl/cmd/CompleteTaskCmd.java [hunk 4]
```diff
@@ -72,6 +73,10 @@ protected Void execute(CommandContext commandContext, TaskEntity task) {
       }
     }
 
+    if(commandContext.getProcessEngineConfiguration().isCopyVariablesToLocalForTasks()){
+      TaskVariableCopier.copyVariablesOutFromTaskLocal(task);
```
[the remark this answers is not in the document]
Comment (ryandawsonuk) [reply activiti-engine/src/main/java/org/activiti/engine/impl/cmd/CompleteTaskCmd.java]: I tried that but from there it is too late as the task has already completed

**Pull request merged**